File: Fitness/Resources/Exercise.m
//
//  Exercise.m
//  Fitness
//
//  Created by Mani on 23.04.15.
//  Copyright (c) 2015 Mani. All rights reserved.
//

#import <ResearchKit/ResearchKit.h>
#import <ResearchKit/ORKCountdownStep.h>
#import <ResearchKit/ORKWalkingTaskStep.h>
#import "Exercise.h"
#import "Training.h"
#import "Helpers.h"

@interface Exercise () <NSCoding, NSCopying, NSMutableCopying>

+ (NSDictionary *)exerciseTypes;

@end

@implementation Exercise

- (id)init {
    self = [super init];
    
    self.type = 0;
    self.warmup = @5;
    self.interval = @60;
    
    return self;
}

#pragma mark NSCoding Protocol
- (id)initWithCoder:(NSCoder *)aDecoder {
    self = [self init];
    
    DECODE_ENUM(self.type, aDecoder);
    DECODE_OBJ(self.warmup, aDecoder);
    DECODE_OBJ(self.interval, aDecoder);
    
    return self;
}

- (void)encodeWithCoder:(NSCoder *)aCoder {
    ENCODE_ENUM(self.type, aCoder);
    ENCODE_OBJ(self.warmup, aCoder);
    ENCODE_OBJ(self.interval, aCoder);
}

#pragma mark NSCopying Protocol
- (id)copyWithZone:(NSZone *)zone {
    Exercise *newObject = [[self class] allocWithZone:zone];
    
    newObject->_type = self.type;
    newObject->_warmup = [self.warmup copyWithZone:zone];
    newObject->_interval = [self.interval copyWithZone:zone];
    
    return newObject;
}

#pragma mark NSMutableCopying Protocol
- (id)mutableCopyWithZone:(NSZone *)zone {
    Exercise *newObject = [[self class] allocWithZone:zone];
    
    newObject->_type = self.type;
    newObject->_warmup = [self.warmup copyWithZone:zone];
    newObject->_interval = [self.interval copyWithZone:zone];
    
    return newObject;
}

#pragma mark Static Methods
+ (NSDictionary *)exerciseTypes {
    return @{
             @(ExerciseTypeRunning): @"Laufen",
             @(ExerciseTypeBreast): @"Brust",
             @(ExerciseTypeBack): @"Rücken",
             @(ExerciseTypeTriceps): @"Trizeps",
             @(ExerciseTypeBiceps): @"Bizeps",
             @(ExerciseTypeLegs): @"Beine",
             @(ExerciseTypeCalves): @"Waden",
             @(ExerciseTypeShoulders): @"Schultern",
             @(ExerciseTypeStomach): @"Bauch",
             @(ExerciseTypeLowerBack): @"Kreuz",
             @(ExerciseTypePause): @"Pause"
    };
}

+ (NSString *)exerciseTypeToString:(enum ExerciseType)type {
    if(type < 0 || type >= ExerciseTypeCount)
        return @"";
    
    return [[self exerciseTypes] objectForKey:@(type)];
}

#pragma mark Dynamic Methods
- (NSArray *)createResearchKitTasks {
    ORKInstructionStep *instructionStep = [[ORKInstructionStep alloc] initWithIdentifier:@"task"];
    instructionStep.title = [[self class] exerciseTypeToString: self.type];
    
    ORKCountdownStep *countdown = [[ORKCountdownStep alloc] initWithIdentifier:@"countdown"];
    countdown.stepDuration = [self.warmup integerValue];
    countdown.shouldPlaySoundOnFinish = YES;
    countdown.shouldSpeakCountDown = YES;
    countdown.title = @"Vorbereitung";
    countdown.text = @"Bereite dich auf die Übung vor";
    
    ORKCountdownStep *training = [[ORKCountdownStep alloc] initWithIdentifier:@"training"];
    training.stepDuration = [self.interval integerValue];
    training.shouldPlaySoundOnFinish = YES;
    training.shouldSpeakCountDown = YES;
    training.title = [[self class] exerciseTypeToString: self.type];
    training.subtitle = @"";
    
    return @[instructionStep, countdown, training];
}

@end

File: Fitness/Core/Exercise.swift
//
//  Exercise.swift
//  Fitness
//
//  Created by Manuel Stampfl on 24.06.15.
//  Copyright (c) 2015 mani1337. All rights reserved.
//

import Foundation

enum ExerciseType: Int {
    case Running = 0, Breast, Back, Triceps, Biceps, Legs, Calves, Shoulders, Stomach, LowerBack, Pause, Count
    
    func localizedName() -> String {
        return NSLocalizedString(self.localizationKey(), comment: "Enumeration to string")
    }
    
    private func localizationKey() -> String {
        switch self {
        case .Running:
            return "EXERCISETYPE_RUNNING"
            
        case .Breast:
            return "EXERCISETYPE_BREAST"
            
        case .Back:
            return "EXERCISETYPE_BACK"
            
        case .Triceps:
            return "EXERCISETYPE_TRICEPS"
            
        case .Biceps:
            return "EXERCISETYPE_BICEPS"
            
        case .Legs:
            return "EXERCISETYPE_LEGS"
            
        case .Calves:
            return "EXERCISETYPE_CALVES"
            
        case .Shoulders:
            return "EXERCISETYPE_SHOULDERS"
            
        case .Stomach:
            return "EXERCISETYPE_STOMACH"
            
        case .LowerBack:
            return "EXERCISETYPE_LOWERBACK"
            
        case .Pause:
            return "EXERCISETYPE_PAUSE"
        default:
            return ""
        }
    }
}

class Exercise: NSObject, NSCoding {
    // MARK: - Internal vars
    var warmup: NSTimeInterval = 5.0
    var duration: NSTimeInterval = 60.0
    var type: ExerciseType = .Running
    
    // MARK: - NSCoding
    @objc required init(coder aDecoder: NSCoder) {
        self.warmup = aDecoder.decodeObjectForKey("warmup") as! NSTimeInterval
        self.duration = aDecoder.decodeObjectForKey("duration") as! NSTimeInterval
        self.type = ExerciseType(rawValue: aDecoder.decodeObjectForKey("type") as! Int)!
    }
    
    @objc func encodeWithCoder(aCoder: NSCoder) {
        aCoder.encodeObject(self.warmup, forKey: "warmup")
        aCoder.encodeObject(self.duration, forKey: "duration")
        aCoder.encodeObject(self.type.rawValue, forKey: "type")
    }
    
    // MARK: - Initializers
    override init() {
    }
}